doc/Requirements.md

```diff
@@ -0,0 +1,526 @@
+   (Why: any criterion can be crated with this one.)
+
+The selection criterion possible states SHOULD be specifiable by:
+ - A combination of values
+   (Note: combination in the mathematical sense “ab” -> [“”, “a”, “b”, “ab”])
+   (Note: called inclusive criterion)
+   (Note: Un empty value set is allowed as it's combination -a set containing the empty set- would not be empty.
+          The empty set would be the only possible criteria state.)
+   (Why: The reference implementation supports it.)
+
+### Definitions
+ - Selection criterion rule: function (in the mathematical sense) that MUST given selection criteria return a Boolean.
+
+ - Rule: an Boolean expression of Selection criterion rules
+   (Note: implementation only allows AND and OR combination)
+
+### Criterion change
+
+The api to change criterion values MUST allow atomicity regarding configuration application.
+Ie. It MUST be possible to change multiple criterion values without triggering a configuration
+application.
+(Why: Two criterion might have an excluding state.
+      If configuration application was triggered after each)
+
+### Rules
+
+One MUST always be able to create a selection criterion rule from a given selection criterion
+Ie: a criteria MUST always have a state that can be matched by a rule.
+(Why: if no rules can be formulated from a criterion, it is useless)
+
+Parameter values change SHOULD be selected by Rules.
+(Why: A rule based inference engine has been chosen based on implementation and configuration ease)
+
+## Domains
+
+### Definition
+Domain: ordered set of configuration, all of which contain the values for the same parameters.
+
+### Requirement
+
+Each configuration SHOULD be in a "domain" (see Definition chapter).
+(Why: Domains are mostly a way to define the priority of "Scenarios" for some parameters.
+      It is not a MUST because this goal could also be achieve with (for example) global configurations and per parameter priority.
+      It is not a MAY because the reference implementation uses domains.)
+
+If multiple configuration are eligible, the first one MUST be applied.
+(Why: If multiple configuration are eligible)
+
+If no configuration is eligible, no configuration MUST be applied.
+(Note: It means that if none of the configurations is eligible, none is applied.
+       This also mean that no function can be defined between criteria and states.
+       Ie: Parameter values MAY depend on previous selection criterion states.
+(Why: This is what the reference implementation does.)
+
+TODO sequence aware domains
+
+# (de)serialization
+
+## Philosophy
+Serialization and deserialization are meant to support destruction recovery
+and configuration deployment.
+
+These are the same requirements than for a database, it needs to be able to save
+its state and restore for backup, deployment, reboot...
+
+## Definition
+PF data includes:
+ - parameter tree
+ - parameter values
+ - configurations
+ - criteria rule
+ - domains
```

Comment:
Parameter values in which context ? What is the content of a domain ?

I'd rewrite this as:
- parameters tree
- configurations:
  - selection rules
  - parameter/value couples
- domain:
  - list of associated configurations
  - order of priority